File: 2025/golang/day5/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type IDWindow struct {
	start int
	end   int
}

func isFresh(availableIds []IDWindow, line int) bool {
	for _, id := range availableIds {
		if line >= id.start && line <= id.end {
			return true
		}
	}
	return false
}

func addAvailableIds(availableIds []IDWindow, line string) []IDWindow {
	windows := strings.Split(line, "-")
	start, err := strconv.Atoi(windows[0])
	if err != nil {
		panic(err)
	}
	end, err := strconv.Atoi(windows[1])
	if err != nil {
		panic(err)
	}
	return append(availableIds, IDWindow{start: start, end: end})
}

func solveDay5Part1(input string) {
	file, err := os.Open(input)
	if err != nil {
		panic(err)
	}

	total := 0
	scanner := bufio.NewScanner(file)
	availableIds := []IDWindow{}
	isSplitted := false
	for scanner.Scan() {
		line := scanner.Text()
		if len(line) == 0 {
			isSplitted = true
			continue
		}

		if isSplitted {
			id, err := strconv.Atoi(line)
			if err != nil {
				panic(err)
			}
			if isFresh(availableIds, id) {
				total++
			}
			continue
		}

		availableIds = addAvailableIds(availableIds, line)

	}

	if err := scanner.Err(); err != nil {
		panic(err)
	}

	fmt.Println("Total: ", total)
}

func main() {
	// input := "input_test.txt" // Expected output: 3
	input := "input.txt" // 737
	solveDay5Part1(input)
}

File: 2025/cpp/day5/part1.cpp
#include <iostream>
#include <fstream>
#include <vector>



typedef struct {
    long long start;
    long long end;
} IDWindow;


int is_fresh(std::vector<IDWindow> &available_ids, long long line) {
    for (auto &available_id: available_ids) {
        if (line >= available_id.start && line <= available_id.end) {
            return 1;
        }
    }
    return 0;
}

void add_available_ids(std::vector<IDWindow> &available_ids, std::string line) {
    long long start;
    long long end;
    size_t del_position = line.find('-');
    start = std::stoll(line.substr(0, del_position));
    end = std::stoll(line.substr(del_position + 1));

    available_ids.push_back(IDWindow{start, end});
}

void solve_day5_part1(std::string input_file) {
    using namespace std;

    int total = 0;
    int is_splitted = 0;
    vector<IDWindow> available_ids;
    ifstream input(input_file);
    string line;

    while (getline(input, line)) {
        if (line.empty()) {
            is_splitted = 1;
            continue;
        }

        if (is_splitted) {
            long long id = stoll(line);
            total += is_fresh(available_ids, id);
            continue;
        }

        add_available_ids(available_ids, line);
    }

    cout << "Total: " << total << endl;
}

int main() {
    // std::string input_file = "input_test.txt"; // Expected output: 3
    std::string input_file = "input.txt"; // 737
    solve_day5_part1(input_file);

    return 0;
}
